### DISCUSSION TITLE
What is preferred way to deal with peerDependencies

### DISCUSSION THREAD

**Discussion by justsaul** [yarnpkg/berry#4791] [Q&A] [answered]
### Hey folks,

I was wondering what is preferred way to deal with workspaces that rely on **peerDependencies**.

Say we have a project - `app` and package - `package`. Both are expressed as a workspaces and `app` depends on package `package` for a particular functionality. Now say `package` has `react` **peerDependency** (which is expected to be provided by an `app`). Now conventional way in order to "build" and "test" `package` is to declare **devDependencies** identical to **peerDependencies**.

Now the problem arises that after linking step, within `app` we literally get symlink to `package` with all its _node_modules_. (similar problem happens with PnP linker as opposed to node-modules where we would get `package` referencing **devDependency** - in this case `react`).

Module bundlers allow us to exclude such **peerDependencies** that happened to became **devDependencies** out of necessity, however, since package still contains all _node_modules_ at runtime `react` **dependency** will will be taken from `package node_modules` even though, it's just a **devDependency** and if `app` happened to have different version of `react` then it will not work properly in local env.

### More Context

`package` can be deployable, so say if there is `app2`, it could depend on `package` version from an artifactory but `app` would still depend on local version of `package`.

### Ideal case

Ideally `app` would have symlink to `package` build dir only or to artefact that we get whilst packing `package`. However, i have to support `app2` that has dependency on `package` from artifactory.

**Comment by justsaul** [comment, marked as answer]
For the future people that may or may not stumble upon this discussion - try to move to PnP mode, my initial assumptions have been incorrect, yarn already knows how to resolve your dependencies correctly and you shouldn't run into cases like mentioned above.

**Reply by akwodkiewicz** [reply]
I was going to ask the exact same question, and this was the first post I saw in the "Discussions" tab. Thank you for this.

Does this mean that PnP "automatically" resolves this problem? How to migrate from node_modules linker to PnP if I already use Yarn 3?
